### PR TITLE
[RISCV] Allow VCIX with SE to reorder

### PR DESCRIPTION
**Pull request by 4vtomat**
This patch allows VCIX instructions that have side effect to be reordered
with memory and other side effecting instructions. However we don't want
VCIX instructions to be reordered with each other, so we propose a dummy
register called VCIX_STATE and make these instructions implicitly define and use
it.

### CODE REVIEW

**Comment by topperc** [comment]
CC: @michalt

**Comment by topperc** [comment]
> This patch sets HasSideEffect to 0, it allows VCIX instructions
that have side effect to be reordered.

This description needs to be improved. We don't want the VCIX instructions to be reordered relative to each other. We just want them to be able to reorder with memory and other side effecting instructions.

Should also mention the VCIX_STATE.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -8832,72 +8901,117 @@ SDValue RISCVTargetLowering::LowerINTRINSIC_VOID(SDValue Op,
         FixedIntrinsic->getMemoryVT(), FixedIntrinsic->getMemOperand());
   }
   case Intrinsic::riscv_sf_vc_x_se_e8mf8:
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:8903]: Would it be possible to redefine riscv_sf_vc_x_se* and riscv_sf_vc_i_se IR intrinsics store the sew and lmul as immediate operands like vsetvli. Then use those immediate operands during isel to select the instructions? That would remove 42 intrinsics I think.

**Comment by michalt** [comment]
Very cool, thanks! 😀

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 1 above)
Comment (4vtomat) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:8903]: If we redefine `riscv_sf_vc_x_se*` to `riscv_sf_vc_x_se`, we should also change C intrinsics to match those IRs, but we are not able to distinguish between C intrinsics if we don't have suffix. We can make sew and luml as input operands in IR, but we can't add those to C intrinsics, can we?

**Comment by topperc** [comment]
> propose a dummy CSR called VCIX_STATE 

nit: it's not a CSR. CSR is a RISC-V term for register accessed by instructions like `csrrw`.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**